core/src/pdf417/PDFDecodedBitStreamParser.cpp

```diff
@@ -571,12 +573,15 @@ DecodeStatus DecodeMacroBlock(const std::vector<int>& codewords, int codeIndex,
 	if (StatusIsError(status)) {
 		return status;
 	}
-
 	resultMetadata.setSegmentIndex(std::stoi(strBuf));
 
-	std::string fileId;
-	codeIndex = TextCompaction(codewords, codeIndex, fileId);
-	resultMetadata.setFileId(fileId);
+	std::ostringstream fileId;
+	fileId.fill('0');
+	for (int i = 0; codeIndex < codewords[0] && codewords[codeIndex] != MACRO_PDF417_TERMINATOR
+			&& codewords[codeIndex] != BEGIN_MACRO_PDF417_OPTIONAL_FIELD; i++, codeIndex++) {
+		fileId << std::setw(3) << codewords[codeIndex];
+	}
+	resultMetadata.setFileId(fileId.str());
```

Comment:
For the record: I stumbled upon this discrepancy between the code and the spec as well. I found at least one other library that does what ZXing did (allow 'random text') and since the id is only really required to be equal within a set, I thought, why not keep the behavior and allow the users to give it some meaning? Now I checked a little further and found other libs that model it as you just did as well as model it as a single integer (which is obviously insufficient).

Question: is your motivation for this change simply to make it 'right (TM)' or do you have a functional requirement?

Note: I just checked the PDR417Writer and there seems to be no support for MacroBlocks at all, so no inconsistencies there... Does zint strictly follow the spec?

Reply:
Yes the motivation is simply to make it 'right (TM)', and try to follow the spec - there's no functional requirement. The random text generated for the spec examples looks pretty weird though.

 (Zint doesn't support structured append for any barcodes, though it's something I want to add at some stage in the next hundred years.)

I'm trying to check out TEC-IT, though only have the demo so will have to massage its output (it accepts text for the id anyway). Will report back....

Reply:
Ok TEC-IT encodes the id using text compaction, so that's one in the "de facto" camp.

Reply:
Ok found a functional reason to follow the spec - depending on the fileId values, e.g. 899, it would decode into shifts and latches etc, dropping the info. Also different fileIds could resolve to the same text encoded value...

Reply:
Fair enough. Would add a comment describing those insights? (difference compared to upstream/TEC-IT and the reason to switch)

Reply:
Done. I plan to do this change upstream too so left that bit out.